Add Delegates on probation and refactor WCAT email

Jambrose777 opened the pull request:
Currently this PR has added a probation team so that seniors can modify the team and upon a competition submission WCAT is notified in the email that is sent. This would take care of https://github.com/thewca/worldcubeassociation.org/issues/3885.

In this PR I have:
- [x] Added a hidden probation team.
- [x] Gave permission for Senior Delegates and WFC Leader/Senior Members to edit the probation team.
- [x] Added a link in the Senior and new WFC Panel to edit the probation team.
- [x] Adjusted the WCAT email to be meet the new requirements here: https://github.com/thewca/worldcubeassociation.org/issues/4143
- [x] Fixed the days to be announced to be 28 instead of 29 warning.
- [x] Added validator that the competition must be at least 28 days in advanced in order to submit (unless they are an admin for competitions).
- [x] modified regulation Z1 and Z3 display text (the variable names are used everywhere, so even though these regulations no longer exist, they still represent what they used to).
- [x] From https://github.com/thewca/worldcubeassociation.org/issues/4143 Add link to Probation sheet.
- [x] Create tests / fix coverage and travis.

Questions:
1. I added probation.name in the en.yml file. For the other languages is there anything I would need to do to notify the translators to update this, or is that something they do whenever they see that this is missing?
2. I have a list of queries to run:
  - The table insertion query: INSERT INTO `teams` (`id`, `friendly_id`, `created_at`, `updated_at`, `email`, `hidden`) VALUES (19, 'probation', '2019-05-14 00:00:00', '2019-05-14 00:00:00', NULL, '1'); This may already be taken care of with the migration script, I'm not positive.
  - A list of team_member inserts created based on the spreadsheet. INSERT INTO `team_members` (`id`, `team_id`, `user_id`, `start_date`, `end_date`, `team_leader`, `created_at`, `updated_at`, `team_senior_member`) VALUES (NULL, '19', '5569', '2019-05-05', NULL, '0', '2019-05-14 00:00:00', '2019-05-14 00:00:00', '0');
  Do I need to add these queries anywhere or is it fine to just run them once this is merged in? Also of note, these queries contain sensitive information with the user_id field, above I just used mine as an example.
3. How do we make the entries in team_members where team_id = 19 not appear in public and dev export?

Items that I will implement in a later PR which aren't fully needed for this to be functioning:
- (started work on this here: https://github.com/Jambrose777/worldcubeassociation.org/tree/adding-probation-reasons-attempt) On the Delegates on Probation Spreadsheet, there are additional fields End Terms, Restrictions, Reason. If we want to completely move off the spreadsheet, we'd also need to add these fields in somehow? For now this would just make it 95% easier for WCAT while making it a tad more inconvenient for Seniors since they would still have to update the sheet and WCAT would still have to access the sheet to understand the delegate's Restrictions. Once implemented the Restrictions would replace "Please check the Delegates on Probation spreadsheet for more details." in the email. But this may be something to add in a separate PR afterwards. Edit- There is also a progress column we can add. (depending on the complexity we can move this out of the PR).

End Terms | Restrictions | Reason
-- | -- | --
3 consecutive competitions with reports submitted on time | Cannot delegate competitions held in the same week or held on consecutive weekends | Too many submissions of late Delegate reports
- give reason for probation in WCAT email.
- Second half of https://github.com/thewca/worldcubeassociation.org/issues/3885 (which can probably be implemented at a later date)

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -4,6 +4,13 @@
 
+<% @competition.delegates.map do |delegate| %> 
+  <% if Team.probation.current_members.map { |member| member.user_id }.include?(delegate.id) %>
+  <p>  </p>
```
The <p> is left over. I'll remove this when modifying the email with the other items.

Jambrose777 commented:
For the extra probationary information. Maybe a table like this?

CREATE TABLE `team_member_details` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `team_member_id` int(11) NOT NULL,
  `reasons` text COLLATE utf8mb4_unicode_ci,
  `restrictions` text COLLATE utf8mb4_unicode_ci,
  `end_terms` text COLLATE utf8mb4_unicode_ci,
  `progress` text COLLATE utf8mb4_unicode_ci,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

Mollerz commented:
Can we also allow WFC Senior Members to also edit probation? Dues are fairly automated to the WFC Leader does not actually see these tasks day in day out. Probation is approved by the senior in charge of Dues (which is currently me).

Jambrose777 commented:
> Can we also allow WFC Senior Members to also edit probation? Dues are fairly automated to the WFC Leader does not actually see these tasks day in day out. Probation is approved by the senior in charge of Dues (which is currently me).

Just added Senior Members to the permissions. The Board also brought this up when I was researching this, so it should be good!

AlbertoPdRF commented on WcaOnRails/db/seeds/teams.seeds.rb:
```diff
@@ -18,3 +18,4 @@
 Team.create(friendly_id: 'vice_chair')
 Team.create(friendly_id: 'wdpc', email: "dataprotection@worldcubeassociation.org")
 Team.create(friendly_id: 'wac', email: "advisory@worldcubeassociation.org")
+Team.create(friendly_id: 'probation', hidden: true)
```
You are missing a blank line at the end of this file.

Jambrose777 replied:
Fixed.

AlbertoPdRF commented on WcaOnRails/db/structure.sql:
```diff
@@ -974,7 +974,7 @@ DROP TABLE IF EXISTS `rails_persons`;
 SET @saved_cs_client     = @@character_set_client;
 SET character_set_client = utf8;
-/*!50001 CREATE VIEW `rails_persons` AS SELECT
+/*!50001 CREATE VIEW `rails_persons` AS SELECT 
```
There's an extra white space at the end of this file.

Jambrose777 replied:
fixed. Though I don't recall adding this. This file was modified when I ran the migration script.

Jambrose777 replied:
and it added back in because I had to run $bin/rails db:migrate RAILS_ENV=development.

Any reason this would do this?

AlbertoPdRF replied:
It gets automatically added for me as well when I run any migration along with the `/*!50013 DEFINER=`root`@`localhost` SQL SECURITY DEFINER */` line :confused: I just make sure to not commit those two anymore since I'm aware of them.

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -269,10 +269,10 @@ en:
-        regulation_z1_reason: "Reason for Z1)"
-        regulation_z3: "I would like to select competitors who directly qualify for certain rounds of certain events based on the results of specific previous competitions as per Regulation Z3)"
-        regulation_z3_reason: "Reason for Z3)"
+        regulation_z1: "I would like to require competitors to submit puzzles at a specific time before an event is scheduled to begin"
```
Shouldn't we rename these fields on the database know that Regulations Zn) don't exist anymore?

Jambrose777 replied:
Probably,
However, regulation_z1_reason and regulation_z3_reason have 48 uses each, regulation_z1 and regulation_z3 have 97 uses each.
This is a bit much for me to refactor at this time without really knowing what I'm doing.

AlbertoPdRF replied:
Many of their uses just come from the translations, so you don't need to bother with those, you just need to change the uses on the actual code and on the English locale file.

Jambrose777 replied:
This would also involve changing it in the database correct?

AlbertoPdRF replied:
Correct.

jfly replied:
There's already a lot of different stuff going on in this PR. I'd vote for renaming these in a separate PR.

AlbertoPdRF replied:
Yep, they will be renamed on a different PR. We already have an issue to track this: #4217.

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -269,10 +269,10 @@ en:
-        regulation_z3_reason: "Reason for Z3)"
+        regulation_z1: "I would like to require competitors to submit puzzles at a specific time before an event is scheduled to begin"
+        regulation_z1_reason: "Reason for submitting puzzles early"
+        regulation_z3: "I would like to select competitors who directly qualify for certain rounds of certain events based on the results of specific previous competitions"
```
Did we move Regulation Z3) to the [WCA Competition Requirements Policy](https://www.worldcubeassociation.org/documents/policies/external/Competition%20Requirements.pdf) or did we get rid of it? I did a quick check and couldn't find it.

Jambrose777 replied:
WCRP 5.1 about qualification times. My understanding is that qualification times are a subset of Z3 and we got rid of the rest of it (ie qualifying for non first-rounds)

AlbertoPdRF replied:
That's my understanding too, so the `regulation_z3` text should definitely be modified.

Jambrose777 replied:
Updated!

Jambrose777 commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -862,6 +862,13 @@ def has_date_errors?
 
+  validate :start_date_must_be_28_days_in_advanced
+  def start_date_must_be_28_days_in_advanced
+    if !User.find(editing_user_id).can_admin_competitions? && days_until < SHOULD_BE_ANNOUNCED_GTE_THIS_MANY_DAYS
```
It allows can admin competitions to edit the competition even if the date is in the past. 

I’ll retest this to make sure I’m correct, but I’m pretty sure it wouldn’t let me edit the competition in the without.

Jambrose777 replied:
Right, so If I, logged in as myself, went to go update the information section of a competition that is less than 28 days in the future, I would get this error if that snippet wasn't there:
![image](https://user-images.githubusercontent.com/11577241/58923050-e5204c80-870b-11e9-895a-cf4488cf8064.png)

AlbertoPdRF replied:
I see. I think what I first exposed is still true and shouldn't, but I don't see now a way to handle that case.

By the way, shouldn't the `SHOULD_BE_ANNOUNCED_GTE_THIS_MANY_DAYS` variable be renamed to `MUST_BE_ANNOUNCED_GTE_THIS_MANY_DAYS` to keep it consistent with the current regulations?

Jambrose777 replied:
Agreed and changed!

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,89 @@
-<h1><%= @competition.name %> is confirmed</h1>
+Competition Announcement Team, <br>
```
Sure!
Here is the main body with everything added:
![image](https://user-images.githubusercontent.com/11577241/58923358-5b717e80-870d-11e9-98d2-0a1fd10b6e2f.png)
The extra logic at the top has to do with the date being multi-day and spanning across multiple months/years.

Jambrose777 replied:
multi-day:
![image](https://user-images.githubusercontent.com/11577241/58923412-a4293780-870d-11e9-941a-65fd1c8240b3.png)
multi-month:
http://localhost:3000/rails/mailers/competitions_mailer/notify_wcat_of_confirmed_competition
multi-year:
![image](https://user-images.githubusercontent.com/11577241/58923447-c9b64100-870d-11e9-8444-a937ec3d4f60.png)

AlbertoPdRF replied:
Thanks! These LGTM.

Jambrose777 commented on WcaOnRails/app/mailers/competitions_mailer.rb:
```diff
@@ -10,11 +10,12 @@ def notify_wcat_of_confirmed_competition(confirmer, competition)
     I18n.with_locale :en do
       @competition = competition
       @confirmer = confirmer
+      @months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
```
select_month gives me a drop down that defaults to that month:
![image](https://user-images.githubusercontent.com/11577241/58923590-4a753d00-870e-11e9-95e4-7a924dbe76ac.png)
However this works:
`Date::MONTHNAMES[@competition.month]`

AlbertoPdRF replied:
Nice! Now it looks cleaner :)

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,89 @@
+For a complete list of confirmed but unannounced competitions see your <%= link_to "notifications", notifications_url %>.<br>
+<br>
+Regards,<br>
+<% @competition.delegates.sort_by { |d| d.name }.each.each_with_index do |delegate, index| %>
```
is it possible to get the 'and' in there like this?

Jambrose777 replied:
.to_sentence is good for this: https://codereview.stackexchange.com/questions/5863/ruby-function-to-join-array-with-commas-and-a-conjunction

I've changed it to that.

AlbertoPdRF replied:
Perfect!

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,89 @@
+Regards,<br>
+<% @competition.delegates.sort_by { |d| d.name }.each.each_with_index do |delegate, index| %>
+  <%= delegate.name %><% if index != @competition.delegates.map.size - 1 %>, <% end %><% if index == @competition.delegates.map.size - 2 %>and <% end %>
+  <% end %>
```
Fixed.

Jambrose777 commented on WcaOnRails/app/views/panel/index.html.erb:
```diff
@@ -44,9 +44,24 @@
 
+
+  <% if current_user.team_leader?(Team.wfc) || current_user.team_senior_member?(Team.wfc) %>
+    <h2>WFC</h2>
```
Fixed the first, but kept the ordering. This order has less permissions at the top down to higher ups:
- Staff
- Delegate
- Seniors
- Team members (aka just WFC here)
- Team leaders
- Board

Jambrose777 replied:
So for Chris, it would look like:
![image](https://user-images.githubusercontent.com/11577241/58923965-d2a81200-870f-11e9-943b-2f8a4b0a9aad.png)

AlbertoPdRF replied:
Ok, makes sense.

Jambrose777 commented on WcaOnRails/db/migrate/20190514234342_add_hidden_team_for_delegates_on_probation.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddHiddenTeamForDelegatesOnProbation < ActiveRecord::Migration[5.1]
+  def change
+    Team.create(friendly_id: 'probation', rank: 100, hidden: true)
```
removed and seems to work fine.

Jambrose777 commented:
Good news, I fixed the tests!
Bad news, I had to comment a test out that was logically incorrect.

So still on my TODOs:
- [ ] add tests for what I changed (should be easier for me now that I've fixed tests and have a better understanding)
- [ ] fix validation for confirmed competitions (shouldn't be able to edit certain fields https://github.com/thewca/worldcubeassociation.org/issues/4201)
- [ ] fix tests for the validation things
- [ ] refactor z3 and z5
- [ ] Still need answers / help to my original 2 questions as well as the second todo item.

I also found a better solution to the messy  date logic with the wca_date_range function, so i use that instead..

Jambrose777 commented:
I attempted to refactor z1 and z3, however this made tons of tests fail due to it not recognizing the renamed method 'regulation_z3?' -> 'qualification_time?' that the method isnt recognized, and I wan't able to figure out where in the code this would be defined.

Jambrose777 commented on WcaOnRails/db/structure.sql:
```diff
@@ -1328,7 +1328,8 @@ CREATE TABLE `wcif_extensions` (
 /*!50001 CREATE ALGORITHM=UNDEFINED */
-/*!50001 VIEW `rails_persons` AS select `Persons`.`rails_id` AS `id`,`Persons`.`id` AS `wca_id`,`Persons`.`subId` AS `subId`,`Persons`.`name` AS `name`,`Persons`.`countryId` AS `countryId`,`Persons`.`gender` AS `gender`,`Persons`.`year` AS `year`,`Persons`.`month` AS `month`,`Persons`.`day` AS `day`,`Persons`.`comments` AS `comments`,`Persons`.`incorrect_wca_id_claim_count` AS `incorrect_wca_id_c ...
+/*!50013 DEFINER=`root`@`localhost` SQL SECURITY DEFINER */
+/*!50001 VIEW `rails_persons` AS select `persons`.`rails_id` AS `id`,`persons`.`id` AS `wca_id`,`persons`.`subId` AS `subId`,`persons`.`name` AS `name`,`persons`.`countryId` AS `countryId`,`persons`.`gender` AS `gender`,`persons`.`year` AS `year`,`persons`.`month` AS `month`,`persons`.`day` AS `day`,`persons`.`comments` AS `comments`,`persons`.`incorrect_wca_id_claim_count` AS `incorrect_wca_id_c ...
```
Tests always failed without this, so im not totally why.

Jambrose777 replied:
and it seems tests fail because of this too.......

Jambrose777 commented:
Would anyone from WST be able to look over this? Alberto wanted a second pair of eyes. It also appears there are merge conflicts again, so I'll fix those once it has been reviewed.

I couldn't figure out the handle to ping software :(
@jfly @Baiqiang @FatBoyXPC @jonatanklosko @lgarron @pedrosino @viroulep @ronaldmansveld

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -269,10 +269,10 @@ en:
-        regulation_z3_reason: "Reason for Z3)"
+        regulation_z1: "I would like to require competitors to submit puzzles at a specific time before an event is scheduled to begin"
+        regulation_z1_reason: "Reason for submitting puzzles early"
+        regulation_z3: "I would like to select competitors who directly qualify for events at this competition based on the results of specific previous competitions"
```
I would re-write this sentence, as it doesn't sound quite good. Also, maybe we should include another field for the optional requirement of forbidding certain combinations of events. Or maybe this last part can go together with #4217.

Jambrose777 replied:
Sure or be its own thing.
I tried to stay away from modifying the form. I only did it for the 28 day thing since that was easy enough and causes WCAT issues, whereas we usually catch the combination of events, so that addition isn't as urgent. I've added it to the WCAT wish list: https://github.com/thewca/worldcubeassociation.org/issues/3759#issuecomment-515283243

Jambrose777 replied:
I adjusted the sentence as well as changed the reasons (to include a description of what is being requested)

AlbertoPdRF replied:
Perfect!

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -269,10 +269,10 @@ en:
+        regulation_z1: "I would like to require competitors to submit puzzles at a specific time before an event is scheduled to begin"
+        regulation_z1_reason: "Reason for submitting puzzles early"
+        regulation_z3: "I would like to select competitors who directly qualify for events at this competition based on the results of specific previous competitions"
+        regulation_z3_reason: "Reason for qualification times"
```
Nit: `times`->`results`.

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -455,9 +455,9 @@ en:
+        regulation_z1_reason: "Please elaborate here why you would like to require competitors to submit puzzles early. Please fill this out in English!"
         regulation_z3: ""
-        regulation_z3_reason: "Please elaborate here why you would like to adopt Regulation Z3). Please fill this out in English!"
+        regulation_z3_reason: "Please elaborate here why you would like to use qualification times. Please fill this out in English!"
```
Nit: `times`->`results`.

AlbertoPdRF commented on WcaOnRails/config/locales/en.yml:
```diff
@@ -1152,6 +1152,7 @@ en:
       must_contain_delegate: "must contain at least one WCA Delegate"
       not_all_delegates: "are not all Delegates"
       registration_close_after_open: "registration close must be after registration open"
+      start_date_must_be_28_days_in_advanced: "Start date must be at least 28 days in the future."
```
Same here re: `advance`

AlbertoPdRF commented on WcaOnRails/db/structure.sql:
```diff
@@ -1007,7 +1007,7 @@ DROP TABLE IF EXISTS `rails_persons`;
 SET @saved_cs_client     = @@character_set_client;
 SET character_set_client = utf8;
-/*!50001 CREATE VIEW `rails_persons` AS SELECT
+/*!50001 CREATE VIEW `rails_persons` AS SELECT 
```
Remember to get rid of the whitespace at the end of this line.

AlbertoPdRF commented on WcaOnRails/db/structure.sql:
```diff
@@ -1328,6 +1328,7 @@ CREATE TABLE `wcif_extensions` (
 /*!50001 SET character_set_results     = utf8mb4 */;
 /*!50001 SET collation_connection      = utf8mb4_unicode_ci */;
 /*!50001 CREATE ALGORITHM=UNDEFINED */
+/*!50013 DEFINER=`root`@`localhost` SQL SECURITY DEFINER */
```
Remember to get rid of this line.

AlbertoPdRF commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -562,6 +563,12 @@
         expect(competition.reload.confirmed?).to eq true
       end
 
+      it "cannot confirm the competition that is not at least 28 days in the future" do
```
Uber nit: `the competition`->`a competition`

AlbertoPdRF commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -552,7 +552,8 @@
       end
 
-      it "can confirm competition and expects board and organizers to receive a notification email" do
+      it "can confirm competition and expects wcat and organizers to receive a notification email" do
```
Uber nit: `a competition`.

AlbertoPdRF commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -589,15 +596,15 @@
 
-      it "can change registration open/close of locked competition" do
-        competition.update_attribute(:confirmed, true)
+      # it "can change registration open/close of locked competition" do
```
As fixing this test doesn't quite fit on this PR, could you please open an issue about it and add a `# FIXME: ` comment linking to the issue?

AlbertoPdRF replied:
This is going to be handled as soon as #4333 is merged!

AlbertoPdRF commented on WcaOnRails/spec/controllers/teams_controller_spec.rb:
```diff
@@ -161,5 +161,17 @@
       end
     end
+
+    context "Senior Delegates managing the probation team, despite not being a member of the probation team" do
```
Could we also add tests for the WFC Leader and a WFC Senior Member?

Besides, I think we never capitalize stuff here.

AlbertoPdRF replied:
We should probably add tests like [this one](https://github.com/thewca/worldcubeassociation.org/pull/4144/files#diff-dadb4202f6cdc5dc7b128619017dc973R63) too.

Jambrose777 replied:
sweet for the capitalization. It seems we do capitalize stuff like WDC, so Ill do the same for WFC and then leave the other words lowercase.

AlbertoPdRF commented on WcaOnRails/spec/controllers/teams_controller_spec.rb:
```diff
@@ -161,5 +161,17 @@
     end
+
+    context "Senior Delegates managing the probation team, despite not being a member of the probation team" do
+      sign_in { FactoryBot.create :user, :senior_delegate }
```
You should use `FactoryBot.create :senior_delegate` here instead.

AlbertoPdRF commented on WcaOnRails/spec/factories/users.rb:
```diff
@@ -170,6 +170,10 @@
       end
     end
 
+    trait :senior_delegate do
```
This is not necessary, as you can simply use `FactoryBot.create :senior_delegate`.

AlbertoPdRF commented on WcaOnRails/spec/requests/competitions_spec.rb:
```diff
@@ -57,6 +57,7 @@
     context "signed in as a delegate" do
       before :each do
         sign_in competition.delegates.first
+        competition.update!(start_date: 5.week.from_now, end_date: 5.week.from_now)
```
Wouldn't it be better to just specify this when creating the competition at the beginning of the file?

Jambrose777 replied:
I guess, but I didn't want to affect any of the admin tests. Let me know if you still want this changed (I haven't included any changes for this yet).

AlbertoPdRF replied:
It shouldn't affect any of the tests, I think, but feel free to leave it as is if you think it's better! By the way, uber nit: `week`->`weeks`

AlbertoPdRF commented:
Regarding your original questions:

> 1. I added probation.name in the en.yml file. For the other languages is there anything I would need to do to notify the translators to update this, or is that something they do whenever they see that this is missing?

You don't have to do anything else, the translators will receive an email letting them know that there are new keys that need to be translated.

> 2. I have a list of queries to run:
> - The table insertion query: INSERT INTO teams (id, friendly_id, created_at, updated_at, email, hidden) VALUES (19, 'probation', '2019-05-14 00:00:00', '2019-05-14 00:00:00', NULL, '1'); This may already be taken care of with the migration script, I'm not positive.
> - A list of team_member inserts created based on the spreadsheet. INSERT INTO team_members (id, team_id, user_id, start_date, end_date, team_leader, created_at, updated_at, team_senior_member) VALUES (NULL, '19', '5569', '2019-05-05', NULL, '0', '2019-05-14 00:00:00', '2019-05-14 00:00:00', '0');
> Do I need to add these queries anywhere or is it fine to just run them once this is merged in? Also of note, these queries contain sensitive information with the user_id field, above I just used mine as an example.

The first one is not needed, it will be taken care of with your code. The second one shouldn't be added anywhere, you can just run it once your changes are live.

> 3. How do we make the entries in team_members where team_id = 19 not appear in public and dev export?

I think there's nothing to be done in this regard, as this was already handled when introducing hidden teams.

AlbertoPdRF commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -185,7 +185,7 @@ class Competition < ApplicationRecord
   MAX_SPAN_DAYS = 6
 
   # https://www.worldcubeassociation.org/regulations/guidelines.html#8a4++
```
We should also update this comment to point to the WCRP.

Jambrose777 commented:
> Here goes a more complete review of the PR. I think it's mostly looking good, thanks for the work @Jambrose777!
> 
> By the way, have you looked at [the existing previews of the `competitions_mailer`](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/spec/mailers/previews/competitions_mailer_preview.rb) and considered adding some?

Yes I have seen this file and used it to help code the WCAT email! The first one is the email that I edited. Is there more that need to be added?

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,83 @@
+  The competition will take place on <%= wca_date_range(@competition.start_date, @competition.end_date) %> in <%= (@competition.start_date - Time.now.to_date).to_i %> days.
+  <% if (@competition.start_date - Time.now.to_date).to_i <= 30 %>
+    There are less than 48 hours remaining for the competition to be announced. @Delegates, please respond to any WCAT concerns as soon as possible to ensure that the competition becomes announced. If the competition becomes less than 28 days away, the WCAT cannot announce it.
+    <% end %>
```
Rip, I did the indentation based off the first one here: https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb

Thought it was strange, but didnt really question it :P. If its just the <%end%> tag Ill fix all those in the next push!

AlbertoPdRF replied:
Hahaha, bad luck :stuck_out_tongue:

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,83 @@
+      This competition has a potential proximity conflict with
+      <%= link_to c.name, admin_edit_competition_url(c) %>
+      which is
+      <% if (@competition.start_date - c.start_date).to_i > 0 %>
```
Agreed!

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,83 @@
-<h1><%= @competition.name %> is confirmed</h1>
+Competition Announcement Team, <br>
```
Sounds good to me!

Jambrose777 commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,83 @@
 
-<p>
+<span>
   @WCAT: You can delete this competition or make it publicly visible <%= link_to "here", admin_edit_competition_url(@competition) %>.
```
I like manage!

Jambrose777 commented:
Everything should be taken care of (minus 1 comment above https://github.com/thewca/worldcubeassociation.org/pull/4144#discussion_r307570534). Let me know what you think!

AlbertoPdRF commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,83 +1,92 @@
+<% end %>
 
 <% @competition.nearby_competitions(Competition::NEARBY_DAYS_WARNING, Competition::NEARBY_DISTANCE_KM_WARNING).each do |c| %>
+  <% days_difference = (@competition.start_date - c.start_date).to_i %>
```
This line (and the one below) should go into the `if` statement, as there's no need for it if it's not going to be used.

AlbertoPdRF commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,83 +1,92 @@
 <% if @competition.regulation_z3? %>
   <span>
-    The Delegate(s) requested the adoption of <b>qualification times</b> because "<%= @competition.regulation_z3_reason %>".<br>
+    The <%= "Delegate".pluralize(@competition.delegates.count) %> requested the adoption of <b>qualification times</b> because "<%= @competition.regulation_z3_reason %>".
```
Nit: `times`->`results`

AlbertoPdRF commented on WcaOnRails/spec/controllers/teams_controller_spec.rb:
```diff
@@ -66,6 +66,27 @@
+    end
+
+    it "leader of WFC can manage the probation team, despite not being a member of the probation team" do
+      sign_in FactoryBot.create :user, :wfc_leader
```
This should be: `sign_in FactoryBot.create :user, :wfc_member, team_leader: true`

AlbertoPdRF commented on WcaOnRails/spec/factories/users.rb:
```diff
@@ -140,6 +140,12 @@
       end
     end
 
+    trait :wfc_leader do
```
This shouldn't go here, as you can use `FactoryBot.create :user, :wfc_member, team_leader: true`.

Jambrose777 commented:
> A few minor comments!
> 
> Also, it seems that you have some merge conflicts, please rebase the whole thing when you can 🙂
> 
> EDIT: at least, this is what I see:
> 
> ![image](https://user-images.githubusercontent.com/26309166/61960508-e4968a80-afc5-11e9-96b1-7bac9c895d05.png)
> 
> It would probably be better to squash everything into one commit.

hmmmm 

![image](https://user-images.githubusercontent.com/11577241/61974526-c07e7d80-afb4-11e9-8e3d-8f066b93ed7c.png)

Jambrose777 commented:
Let me know if it looks good now, I rebased and squashed the commits.

viroulep commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,91 @@
+  <span class="alert">
+    This competition is marked as <%= @competition.championships.sort.map(&:name).to_sentence %>.
+    <br>
+    <br>
```
There are a lot of these double `<br>` tags in the email template.
Have you considered using `p` (which introduces a bit more spacing) or adding an inline css to make all `span`/`p` have a little more margin bottom?

You also seem to wrap whole paragraph within `span`s, whereas semantically it's supposed to be used to group inline elements together.
Alternatively `p` is the element which defines a paragraph.
While the visual result may be similar, the meaning attached to both constructs is different.

Jambrose777 replied:
That makes a lot of sense! I'll change this over once the other comment I left has an answer on what it should be changed :D

Jambrose777 replied:
Just pushed a change for this. Let me know if that looks better. I still used <br> in the one paragraph with many lines, but I think based on what you wrote that should be fine. Just let me know!

viroulep replied:
Looks great thanks!

Jambrose777 commented on WcaOnRails/app/models/user.rb:
```diff
@@ -569,7 +569,10 @@ def can_edit_team?(team)
+      (team == Team.banned && team_leader?(Team.wdc)) ||
+
+      # Senior Delegates and WFC Leader and Senior Members can edit Delegates on probation
+      (team == Team.probation && (senior_delegate? || team_leader?(Team.wfc) || team_senior_member?(Team.wfc)))
```
its also the same statement as L568 and L569. I don't think the extra tab is correct given what above L572. I can get behind removing the extra spaces in L573 and L570 if you want.

viroulep replied:
Ah I didn't notice that, then indeed my suggestion is just wrong!

I've taken a look at the file on its own (to have a few more context line than on this review tool), and I agree with your suggestion: the `can_edit_team?` would look better (to me) just by removing the blank lines between the different parts of the main `||` statement :)

Jambrose777 replied:
Sweet, should be good now.

viroulep commented on WcaOnRails/app/assets/stylesheets/email.css.scss:
```diff
@@ -4,6 +4,11 @@ h2.alert {
+  font-weight: bold;
+}
+
 span.alert {
```
Here instead of creating two rules with the same styles, you can just have the same behavior using:
```css
p.alert,
span.alert {
// the styles
}
```

and avoiding the code duplication :)

Jambrose777 replied:
That makes a lot of sense!

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -863,7 +863,18 @@ def days_until
 
   def has_date_errors?
     valid?
```
Any idea what this line of code is doing? Is this just to trigger some `before_validation` callbacks?

Jambrose777 replied:
not entirely sure, but this is what triggers the competition validation error on submission (note you can still create the competition just fine with a date < 28 days):
![image](https://user-images.githubusercontent.com/11577241/62091262-144fc780-b23e-11e9-8fce-304b9547c46f.png)
The last validation error seen here.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -863,7 +863,18 @@ def days_until
+  validate :start_date_must_be_28_days_in_advance, if: :confirmed_or_visible?
+  def start_date_must_be_28_days_in_advance
+    if editing_user_id
+      editing_user = User.find(editing_user_id)
```
This doesn't actually feel like the right way to use rails validations here: ideally all of our validations are written in a way such that they would pass on every row of every table in our database. This looks like it would cause all competitions to become invalid once we're within 28 days of them. This feels more like it should be a `can_confirm_competition?` permission in the users table. That would also solve the weirdness that @viroulep mentions about the `editing_user_id` field.

Unrelated, but is this code accurate with regards to timezones?

jfly commented on WcaOnRails/app/views/competitions_mailer/notify_wcat_of_confirmed_competition.html.erb:
```diff
@@ -1,34 +1,78 @@
+<p class="<%= 'alert' if (@competition.start_date - Time.now.to_date).to_i <= 30 %>">
+  The competition will take place on <%= wca_date_range(@competition.start_date, @competition.end_date) %> in <%= (@competition.start_date - Time.now.to_date).to_i %> days.
+  <% if (@competition.start_date - Time.now.to_date).to_i <= 30 %>
+    There are less than 48 hours remaining for the competition to be announced. @<%= "Delegate".pluralize(@competition.delegates.count) %>: please respond to any WCAT concerns as soon as possible to ensure that the competition becomes announced. If the competition becomes less than 28 days away, the WCAT cannot announce it.
```
How accurate is this statement with regards to timezones?

Jambrose777 replied:
I actually have no clue. 

How it should act is in relation to 00:00 UTC of that day, but I'm not sure how to test that.

Jambrose777 replied:
Added .utc to the time.now, so this now works as intended. Same with above.

jfly commented on WcaOnRails/app/views/panel/index.html.erb:
```diff
@@ -47,9 +47,24 @@
   <% end %>
 
+
+  <% if current_user.team_leader?(Team.wfc) || current_user.team_senior_member?(Team.wfc) %>
```
This duplicates the logic you added to `can_edit_team?` above. Did you intentionally leave Senior Delegates out here? How about changing this to `current_user&.can_edit_team?(Team.probation)`?

Jambrose777 replied:
Yes! The panel is split of into sections of position. So Senior Delegates have a section and WFC has a section. 

So this would appear twice if someone was a Senior Delegate and a WFC Senior Member.

I'd prefer to keep it how it is as its what I think Alberto or Bob also agreed with.

Jambrose777 replied:
If the WFC section would ever have more added to it, then the if statment would become for all WFC members and then the subsection would be based on if they can edit the probationary team. However since for regular WFC members, it would be blank, I think the way it is now is best.

AlbertoPdRF commented:
I've just deployed this to staging! @Jambrose777 can you try if everything works as expected? :slightly_smiling_face:

Jambrose777 commented:
Looks great to me!

AlbertoPdRF commented:
Nice! Thanks for all the great work on this :smile:
